PianoFromAbove/GameState.cpp

```diff
@@ -1105,18 +1064,39 @@ GameState::GameError MainScreen::Logic( void )
         {
             MIDIChannelEvent *pEvent = m_vEvents[m_iStartPos];
             if (pEvent->GetChannelEventType() != MIDIChannelEvent::NoteOn) {
-                if (pEvent->GetChannelEventType() == MIDIChannelEvent::ProgramChange && pEvent->GetChannel() != MIDI::Drums && config.m_bPianoOverride)
-                    pEvent->SetParam1(0);
                 if (pEvent->GetChannelEventType() == MIDIChannelEvent::PitchBend) {
-                    //m_pBends[pEvent->GetChannel()] = (short)(((pEvent->GetParam2() << 7) | pEvent->GetParam1()) - 8192);
-                    m_pBends[pEvent->GetChannel()] = (notex_table[1] - notex_table[0]) * (((short)(((pEvent->GetParam2() << 7) | pEvent->GetParam1()) - 8192)) / (8192.0f / 12.0f));
+                    m_pBends[pEvent->GetChannel()] = (notex_table[1] - notex_table[0]) * (((short)(((pEvent->GetParam2() << 7) | pEvent->GetParam1()) - 8192)) / (8192.0f / m_pBendsRange[pEvent->GetChannel()]));
+                } else if (pEvent->GetEventCode() == 176) {
+                    if (pEvent->GetParam1() == 101 && pEvent->GetParam2() == 0) {
+                        m_RPN_MSB[pEvent->GetChannel()] = true;
+                    } else if (pEvent->GetParam1() == 100 && pEvent->GetParam2() == 0) {
+                        if (m_RPN_MSB[pEvent->GetChannel()]) {
+                            m_RPN_LSB[pEvent->GetChannel()] = true;
+                        }
+                        else {
+                            m_RPN_MSB[pEvent->GetChannel()] = false;
+                            m_RPN_LSB[pEvent->GetChannel()] = false;
+                        }
+                    } else if (pEvent->GetParam1() == 6) {
+                        if (m_RPN_MSB[pEvent->GetChannel()] && m_RPN_LSB[pEvent->GetChannel()]) {
+                            m_pBendsRange[pEvent->GetChannel()] = pEvent->GetParam2();
+                        }
+                        else {
+                            m_RPN_MSB[pEvent->GetChannel()] = false;
+                            m_RPN_LSB[pEvent->GetChannel()] = false;
+                        }
+                    }
+                }
```

Comment:
This doesn't handle seeking. Also, it might be a good idea to add the RPN event codes to ChannelEventType to reduce the amount of "magic numbers" in this block.